src/components/icon/icon-registry.ts

```diff
@@ -0,0 +1,239 @@
+import {Injectable} from 'angular2/core';
+import {Http} from 'angular2/http';
+import {Observable} from 'rxjs/Rx';
+
+/**
+  *  Configuration for an icon, including the URL and possibly the cached SVG element.
+  */
+class IconConfig {
+  svgElement: SVGElement = null;
+  constructor(public url: string) {
+  }
+}
+
+@Injectable()
+export class MdIconRegistry {
+  // IconConfig objects and cached SVG elements for individual icons.
+  // First level of cache is namespace (which is the empty string if not specified).
+  // Second level is the icon name within the namespace.
+  private _iconConfigs = new Map<string, Map<string, IconConfig>>();
+
+  // IconConfig objects and cached SVG elements for icon sets.
+  // Multiple icon sets can be registered under the same namespace.
+  private _iconSetConfigs = new Map<string, IconConfig[]>();
+
+  // Cache for icons loaded by direct URLs.
+  private _cachedIconsByUrl = new Map<string, SVGElement>();
+
+  private _fontClassNamesByAlias = new Map<string, string>();
+
+  private _inProgressUrlFetches = new Map<string, Observable<string>>();
+
+  private _defaultFontSetClass = 'material-icons';
+
+  constructor(private _http: Http) {
+  }
+
+  public addIcon(iconName: string, url: string): this {
```

Comment:
If it's SVG-only, then I think that would make the most sense.  If it supports other formats (ie. PNG), then the more generic option works.